### PR TITLE
core: allow to create long dumps for benchmark

### PR DESCRIPTION
**Pull request by fyrchik**
### Problem
It is inconvenient to benchmark mainnet/testnet dump restore because of 2 reasons:
1. They take quite a long time to restore.
2. To get statistically significant differences in running time, full dump needs to be restored.

### Solution
Allow to create dumps for privatenet which contains "meaningful" (i.e. not dummy) transactions and provides necessary load.
Restore them with `./neogo db restore --unittest --in ./chain.acc`.
Profiling with pprof is also easier, because max load is being constantly applied.

In future we can make it more configurable (e.g. optimize for `transfer` speed etc.).
It should be easy to port this to neo2 if needed.

For example for #1553 even for default settings (5000 blocks, 100 tx per block) time varies significantly:
master: `./neogo db restore --unittest --in ./chain.acc  55,27s user 1,16s system 182% cpu 30,935 total`
1553 (no refcount): `./neogo db restore --unittest --in ./chain.acc  93,18s user 1,48s system 179% cpu 52,602 total`
1553 (with refcount): `./neogo db restore --unittest --in ./chain.acc  100,37s user 1,21s system 177% cpu 57,142 total`

### CODE REVIEW

**Comment by fyrchik** [comment]
Results in header are for `inmemory` db btw. `leveldb` needs to be enabled separately (as also `pprof` and `VerifyBlocks: false`).

**Comment by roman-khimov** [comment]
It seems to be very specific for MPT benching, not sure it's worth a test of its own. Maybe we can create a proper `Benchmark` for this?

File: pkg/core/helper_test.go [hunk 1]
```diff
@@ -142,6 +146,116 @@ func newDumbBlock() *block.Block {
...
+		Account:          neoOwner,
+		Scopes:           transaction.CalledByEntry,
+		AllowedContracts: nil,
+		AllowedGroups:    nil,
+	}}
+	require.NoError(t, signTx(bc, tx))
+	return tx
+}
+
+func dumpChain(t *testing.T, bc *Blockchain, path string) {
```
Comment (AnnaShaleva) [inline pkg/core/helper_test.go:237]: We have a very similar piece of code in cli, perhaps we can make a separate helper with this method? It also would be helpful for neo-bench.

**Review by AnnaShaleva — APPROVED** [review]

**Comment by fyrchik** [comment]
@roman-khimov yes, benchmark is more appropriate. I think separate dump is more convenient because:
1. We need to provide the same dump to each execution, which is more difficult to implement in bench, if we don't store entire chain in memory.
2. Tx generating/signing stage will inflate running time and be shown in pprof.
3. It can be useful for other purposes (bench vs C# node, generate dump for CLI test).

We may move this to `scripts/`, though.

File: pkg/core/helper_test.go (hunk 1 above)
Comment (fyrchik) [reply pkg/core/helper_test.go:237]: May be something like `Blockchain.DumpTo(filename string)`.

Comment (fyrchik) [reply pkg/core/helper_test.go:237]: An when combined with `RestoreFrom` this can be useful for tests/benchmarks.

**Comment by roman-khimov** [comment]
> We may move this to scripts/, though.

Let's do that. I'm fine with refactoring some common test and cli code to reduce duplication, but the logic itself is probably better suited for this category, it's not a test, it's not a benchmark of its own (more of a preparation for it), but it's some useful script that generates things that could be used for various purposes.

File: pkg/core/dump.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package core
```
Comment (roman-khimov) [inline pkg/core/dump.go:1]: Please move it out of `core`.

File: pkg/core/helper_test.go [hunk 3]
```diff
@@ -411,24 +384,7 @@ func addSigners(txs ...*transaction.Transaction) {
 }
 
 func signTx(bc *Blockchain, txs ...*transaction.Transaction) error {
```
Comment (roman-khimov) [inline pkg/core/helper_test.go:396]: Why keeping this wrapper?

File: pkg/core/dump.go [hunk 4]
```diff
@@ -24,3 +28,55 @@ func Dump(bc *Blockchain, w *io.BinWriter, start, count uint32) error {
 	}
 	return nil
 }
+
+// Restore restores blocks from provided reader.
+// f is called after addition of every block.
+func Restore(bc *Blockchain, r *io.BinReader, skip, count uint32, f func(b *block.Block) error) error {
```
Comment (roman-khimov) [inline pkg/core/dump.go:34]: This also shouldn't be a part of the `core`.

**Review by roman-khimov — CHANGES_REQUESTED** [review]

File: pkg/core/helper_test.go (hunk 3 above)
Comment (fyrchik) [reply pkg/core/helper_test.go:396]: Forgot to replace, fixed.

File: pkg/core/dump.go (hunk 4 above)
Comment (fyrchik) [reply pkg/core/dump.go:34]: Fixed.

File: pkg/core/dump.go (hunk 2 above)
Comment (fyrchik) [reply pkg/core/dump.go]: Fixed.

File: cli/server/server.go [hunk 5]
```diff
@@ -10,6 +10,7 @@ import (
 	"github.com/nspcc-dev/neo-go/pkg/config"
 	"github.com/nspcc-dev/neo-go/pkg/core"
 	"github.com/nspcc-dev/neo-go/pkg/core/block"
+	"github.com/nspcc-dev/neo-go/pkg/core/blockchainer/serialize"
```
Comment (roman-khimov) [inline cli/server/server.go:13]: `core/chaindump`?

Comment (fyrchik) [reply cli/server/server.go:13]: I think linter will complain about `chaindump.Dump`, but let's see.

**Comment by roman-khimov** [comment]
Rebase, please.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**